### ISSUE TITLE
Using `Application.Start` and/or `Application.Connect` is better in a code generator

### ISSUE THREAD

**Issue by vasily-v-ryabov** [pywinauto/SWAPY#45]
When clicking on a top level window of an application SWAPY might suggest more menu items:
```
"Application.Start"
"Application.Connect"
----- (separator)
"Click"
... (other wrapper's methods)
```

It's not so hard to retrieve a process info using a top level window handle. Expected code for `Start` method:
```python
app = Application.Start("user_software.exe") # even command line params could be retrieved
```
Preferable methods for `Connect` are listed by priority:
```python
app = Application.Connect(path="user_software.exe")
app = Application.Connect(title="Software Title")
app = Application.Connect(pid=5349)
```
We can let user choose
which kind of `Connect` to use. :)

Like so:
```
"Application.Start"
"Application.Connect (by exe name)"
"Application.Connect (by title)"
"Application.Connect (by process id)"
----- (separator)
"Click"
... (other wrapper's methods)
```

I didn't say about `launcher->worker` case detection, but we may think about that as well.

**Comment by moden-py** [comment]
@vasily-v-ryabov, I'm afraid I do not understand what you mean about `launcher->worker`. Please provide more details.
Thanks.

Adding some usability/functional propositions. At first, just wanted to make the screenshots, but perhaps it will be easier to try the mockup.

So you can check all the propositions in dynamic, and perhaps even change something.
I'm not sure we need all now, just gather all the ideas of the subject together.

List of proposals by priorities:
1. Collect all the related items in a submenu.
```
"Application.Start" ==> ...
"Application.Connect" ==>
        "Application.Connect (by exe name)"
        "Application.Connect (by title)"
        "Application.Connect (by process id)"
----- (separator)
"Click"
... (other wrapper's methods)
```
Such an approach would save the submenu size and makes it easy to add new elements (group of elements).
2. Add `*Custom` action for each group. This will act an appropriate dialog with all parameters.
For example add to the group `Application.Connect` new item - `Custom`. It would open a dialog with all the `Connect` arguments(already filled, if possible). It may be much more easy than changing something in the source code(browse file for Start action, etc.).
3. Root element - PC name, has only the `Custom` actions for all groups, to make possible `Start\Connect` start a completely new app or connect with custom parameters. 

[The mockup](https://github.com/moden-py/SWAPY/blob/dev/new_menu.py)

**Comment by vasily-v-ryabov** [comment]
`launcher->worker` example is `C:\Python34\Lib\idlelib\idle.bat` and `pythonw.exe` process (we work with the worker process in SWAPY). But how about launcher detection? It may be impossible always or sometimes. But let's take a look for the possibility.

**Comment by moden-py** [comment]
I am sure the `cmdline` of the top process should be enough. Of course, a `luncher` may manipulate some `environment vars` before run a `worker`, but hope it is a pretty rare case. Even in this case a user can manually start the `luncher`.
So my algo is:
1. Walk the `worker` parents.
2. Get `cmdline` of the first(top) parent process.

[psutil](https://github.com/giampaolo/psutil) module has `Process` class with needed `parent()` and `cmdline()` methods.

**Comment by airelil** [comment]
Why to put so many options? Just generate the right *connect* string with all options that can be reliably detected:
```python
Application.Connect (path="some.exe", title="some title", pid="1234")
```
While in the menu you don't show the long string but just write: "Application.Connect".

Regarding *launcher-worker*, sometimes the launcher even doesn't exist after launching a *worker*. I wouldn't try too much here, may be just an additional option: *star-n-connect*. It would generate some code that user still has to adjust:
```python
Application.Start(path="launcher.exe")
Application.Connect(title="worker") 
```
I also would be careful with PID generally as it constantly changes.

**Comment by vasily-v-ryabov** [comment]
Agree with @airelil. SWAPY just generates what it knows. Advanced customization of the code should be done in Python IDE I believe.

OK, probably launcher detection is too much for SWAPY. It was just a raw idea.

**Comment by moden-py** [comment]
I prefer the easies way too. Let's summarize:
1. Root element has no items.
2. Top level window has the next popupmenu's items:
 ```
"Application.Start" 
"Application.Connect" 
----- (separator)
"Click"
... (other wrapper's methods)
```
(Without submenus)

The editor already has `app = pywinauto.application.Application()` line.
`Application.Start` produces `app.Start_('worker_cmd_line')` code.
`Application.Connect` produces `app.Connect_(title='worker')` code.

Please note the underscore in the method names. According to the comments `Start` and `Connect` are deprecated.

**Comment by vasily-v-ryabov** [comment]
`Application.Connect` can produce `app.Connect_(path='worker.exe')` or `app.Connect_(title='caption')`. Agree with @airelil that pid is not a reliable stuff.

**Comment by moden-py** [comment]
So which one do? Or give the option to choose?
```
"Application.Start" 
"Application.Connect by path" 
"Application.Connect by title" 
----- (separator)
"Click"
... (other wrapper's methods)
```

**Comment by airelil** [comment]
I would stick to the title. BTW, how big is the menu ? It may not fit to some 15"/14" laptop screens.

**Comment by moden-py** [comment]
With new two items it will be 500pixels high.
Let's start with +
```
"Application.Start" 
"Application.Connect"  (by title)
```
Then add more if needed

**Comment by vasily-v-ryabov** [comment]
OK, that's good starting point.

**Comment by moden-py** [comment]
According to the questions in Stack Overflow, `Start` is more interesting than `Continue`. Switch to `Start` by default.
Also "Application.Start"/"Application.Connect" menu items should update the method on fly.

**Comment by moden-py** [comment]
resolved by a8b85dd0cf4eb6e18d4824eb6fd0dc6b302b0637
